torchmetrics/metric.py

```diff
@@ -377,7 +377,19 @@ def wrapped_func(*args: Any, **kwargs: Any) -> None:
             self._computed = None
             self._update_count += 1
             with torch.set_grad_enabled(self._enable_grad):
-                update(*args, **kwargs)
+                try:
+                    update(*args, **kwargs)
+                except RuntimeError as err:
+                    if "Expected all tensors to be on" in str(err):
+                        raise RuntimeError(
+                            "Encountered different devices in metric calculation.\n"
+                            "This could be due to the metric class not being on the same device as input.\n"
+                            f"Instead of `metric={self.__class__.__name__}(...)` try to do"
+                            f" `metric={self.__class__.__name__}(...).to(device)` where"
+                            " device corresponds to the device of the input."
```

Comment:
can we still include the old error message here to give the actual devices to see where the mismatch is from?

Reply:
done :]

Reply:
@justusschock In a previous comment, I advocated for removing this because it's already shown by the `raise ... from ...` construct. Now it's duplicated:


```python
Traceback (most recent call last):
  File "/home/carmocca/git/pytorch-lightning/kkperf.py", line 5, in <module>
    raise RuntimeError("Expected all tensors to be on")
RuntimeError: Expected all tensors to be on

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/carmocca/git/pytorch-lightning/kkperf.py", line 8, in <module>
    raise RuntimeError(
RuntimeError: Expected all tensors to be on. 
Encountered different devices in metric calculation.
This could be due to the metric class not being on the same device as input.
Instead of `metric=FooMetric(...)` try to do `metric=FooMetric(...).to(device)` where device corresponds to the device of the input.
```

Reply:
But with the stacktrace it doesn't show the actual devices, does it?

Reply:
It does:

```python
name = "FooMetric"
import torch

if __name__ == "__main__":
    try:
        torch.tensor(1, device="cuda:1") +  torch.tensor(1, device="cuda:0")
    except RuntimeError as err:
        if "Expected all tensors to be on" in str(err):
            raise RuntimeError(
                "Encountered different devices in metric calculation.\n"
                "This could be due to the metric class not being on the same device as input.\n"
                f"Instead of `metric={name}(...)` try to do"
                f" `metric={name}(...).to(device)` where"
                " device corresponds to the device of the input."
            ) from err
        raise
```

```python
Traceback (most recent call last):
  File "pl_examples/bug_report/bug_report_model.py", line 6, in <module>
    torch.tensor(1, device="cuda:1") +  torch.tensor(1, device="cuda:0")
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:1 and cuda:0!

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "pl_examples/bug_report/bug_report_model.py", line 9, in <module>
    raise RuntimeError(
RuntimeError: Encountered different devices in metric calculation.
This could be due to the metric class not being on the same device as input.
Instead of `metric=FooMetric(...)` try to do `metric=FooMetric(...).to(device)` where device corresponds to the device of the input.
```